src/redisearch_rs/inverted_index/src/numeric/mod.rs

```diff
@@ -0,0 +1,661 @@
+            }
+            HeaderType::Float32Negative => {
+                packed |= Numeric::FLOAT_TYPE << 3; // 2 bits for type
+                packed |= FLOAT32_NEGATIVE << 5; // 3 bits for small negative float
+            }
+            HeaderType::Float64Positive => {
+                packed |= Numeric::FLOAT_TYPE << 3; // 2 bits for type
+                packed |= FLOAT64_POSITIVE << 5; // 3 bits for big float
+            }
+            HeaderType::Float64Negative => {
+                packed |= Numeric::FLOAT_TYPE << 3; // 2 bits for type
+                packed |= FLOAT64_NEGATIVE << 5; // 3 bits for big negative float
+            }
+            HeaderType::FloatInfinite => {
+                packed |= Numeric::FLOAT_TYPE << 3; // 2 bits for type
+                packed |= FLOAT_INFINITE << 5; // 3 bits for infinite
+            }
+            HeaderType::FloatNegInfinite => {
+                packed |= Numeric::FLOAT_TYPE << 3; // 2 bits for type
+                packed |= FLOAT_NEGATIVE_INFINITE << 5; // 3 bits for negative infinite
+            }
+        }
+
+        [packed]
+    }
+}
+
+/// Writes all slices in `bufs` to the writer, advancing the slices as they are written.
+#[inline(always)]
+fn write_all_vectored<const N: usize, W: Write>(
+    writer: &mut W,
+    mut bufs: [IoSlice<'_>; N],
+) -> std::io::Result<usize> {
+    let total_len = bufs.iter().map(|b| b.len()).sum();
+
+    // In theory we only need the code in the `Ok(n)` branch. However, that performs slow when
+    // the buffers being written is small (less than 13 bytes). Using a profiler shows that the
+    // `write_vectored` call inside this `OK(n)` branch behaves differently than the one on
+    // this match (next line) for some reason which is not clear currently.
+    match writer.write_vectored(&bufs) {
+        Ok(n) if n == total_len => return Ok(n),
+        Ok(0) => {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::UnexpectedEof,
+                "failed to fill whole buffer",
+            ));
+        }
+        Ok(n) => {
+            // Partial write, fall back to loop
+            let mut bufs = bufs.as_mut_slice();
+            IoSlice::advance_slices(&mut bufs, n);
+
+            // Could not write everything in one go, fall back to loop
+            while !bufs.is_empty() {
+                match writer.write_vectored(bufs) {
+                    Ok(0) => {
+                        return Err(std::io::Error::new(
+                            std::io::ErrorKind::UnexpectedEof,
+                            "failed to fill whole buffer",
+                        ));
+                    }
+                    Ok(n) => IoSlice::advance_slices(&mut bufs, n),
+                    Err(e) => return Err(e),
+                }
+            }
+        }
+        Err(e) => return Err(e),
+    }
+
+    Ok(total_len)
+}
+
+#[cfg(test)]
+mod tests;
```

Comment:
Move to a top-level `tests` folder, since we don't rely on private APIs in tests.